### DISCUSSION TITLE
Avoid XSS attacks with Wave apps

### DISCUSSION THREAD

**Discussion by dulajra** [h2oai/wave#2237] [Q&A]
XSS attacks is a common security vulnerability when it comes to web applications. I would like to know due to the architecture of Wave are XSS attacks possible for wave apps? Is there any recommended approaches or best practices to avoid such attacks for Wave apps with input sanitisation?

**Comment by mturoci** [comment]
Every web application that renders dynamic data is vulnerable to XSS. Wave is no exception. Common ways to combat XSS are:

* Set up a strict [CSP](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP). Wave allows for specifying extra headers via `H2O_WAVE_HTTP_HEADERS_FILE`.
* Sanitize user-provided input.

**Reply by dulajra** [reply]
@mturoci What is the format of the file we need to set for `H2O_WAVE_HTTP_HEADERS_FILE`? I set the path to an `.env` file with the following content but getting the following error. 

```
panic: failed parsing HTTP headers from file: EOF
goroutine 1 [running]:
main.main()
	/home/runner/work/wave/wave/cmd/wave/main.go:163 +0x1746
Cannot connect to Wave server, retrying...
Cannot connect to Wave server, retrying...
```

Here is the `.env` file 
```
Content-Security-Policy:"default-src 'self';"

```

**Reply by dulajra** [reply]
Where do we have to place the `.env` file? When I placed it on root of the project and pass as `H2O_WAVE_HTTP_HEADERS_FILE=./.env` for `wave run` then I get `panic: failed reading HTTP headers file: open ./.env: no such file or directory
`. I tried placing the `.env` file on the `www` folder also but still getting the same error.

**Reply by mturoci** [reply]
https://wave.h2o.ai/docs/security/#additional-http-response-headers

**Reply by mturoci** [reply]
The .env file is only used for environment variables. Not headers. As the name suggests, you need to create a file containing the headers file and set its path as a value to `H2O_WAVE_HTTP_HEADERS_FILE` conf option.

**Reply by dulajra** [reply]
@mturoci I passed the same file mentioned in the docs but still I get `panic: failed parsing HTTP headers from file: EOF`

**Reply by dulajra** [reply]
> The .env file is only used for environment variables. Not headers. As the name suggests, you need to create a file containing the headers file and set its path as a value to `H2O_WAVE_HTTP_HEADERS_FILE` conf option.

Yes I get it. It's just the name of the file. Apologies for the confusion. My question is do we need to set the absolute path to the headers file? Can't we set the relative path? I placed the `headers.txt` file in the root project and set `H2O_WAVE_HTTP_HEADERS_FILE=./headers.txt` but still I get `panic: failed reading HTTP headers file: open ./headers.txt: no such file or directory`

**Reply by dulajra** [reply]
@mturoci It worked after setting 2 new lines at the end of the headers file even though it is mentioned as `(make sure there's an empty line at the end)` in the [docs](https://wave.h2o.ai/docs/security#additional-http-response-headers)

**Reply by mturoci** [reply]
> My question is do we need to set the absolute path to the headers file? Can't we set the relative path?

I would recommend using absolute paths in general as they are more predictable and stable. Relative path should work as well, and relative in this case means "relative to the waved" location.

> It worked after setting 2 new lines at the end of the headers file even though it is mentioned as (make sure there's an empty line at the end) in the [docs](https://wave.h2o.ai/docs/security#additional-http-response-headers)

Thanks for pointing this out! Fixed in https://github.com/h2oai/wave/commit/8cd6f8427c7379f9558890c6ce87f88a22d42488.

**Comment by dulajra** [comment]
@mturoci I tried setting the `Content-Security-Policy` headers and it looks like we have to enable `unsafe-inline` for `script-src` and `style-src` because wave needs both to work. Because the inline content is generated internally by the framework I don't think we can use more restrictive sources like `hash-*` or `nonce-*`. Do you have better suggestions that allowing `unsafe-inline`?

**Reply by mturoci** [reply]
> because wave needs both to work

This depends on the features you are using IIRC. Switching from inlinine scripts/styles to separate `.js`/ `.css` files should allow you to set the `unsafe-inline`.

You can see what Wave needs by making a sample hello world app with the desired CSP. Rules that fail on sample Wave app need to be disabled at the moment, but I am open to review the needed rules and see what can be done on Wave side.
